scarlet/blend.py

```diff
@@ -94,90 +95,105 @@ def fit(self, max_iter=200, e_rel=1e-3, min_iter=1, random_skip=0, **alg_kwargs)
+                grad_logL = lambda *X: expand_grads(*X, func=grad_logL_func)
+
+                # same for prior. easier her bc we call them independently
+                grad_logP = lambda *X: tuple(
+                    x.prior(x.view(np.ndarray))
+                    if x.prior is not None and not x.fixed
+                    else 0
+                    for x in X
+                )
+
+                # combine for log posterior
+                _grad = lambda *X: tuple(
+                    l + p for l, p in zip(grad_logL(*X), grad_logP(*X))
+                )
+
+                # step sizes, allow for random skipping of parameters
+                _step = lambda *X, it: tuple(
+                    x.step(x, it=it) if hasattr(x.step, "__call__") else x.step
+                    for x in X
+                )
+                _prox = tuple(x.constraint for x in X)
+
+                # good defaults for adaprox
+                scheme = alg_kwargs.pop("scheme", "amsgrad")
+                prox_max_iter = alg_kwargs.pop("prox_max_iter", 10)
+                callback = partial(
+                    self._callback,
+                    e_rel=e_rel,
+                    callback=alg_kwargs.pop("callback", None),
+                    min_iter=min_iter,
+                )
+
+                # do we have a current state of the optimizer to warm start?
+                for x in X:
+                    if x.m is None:
+                        x.m = np.zeros(x.shape)
+                    if x.v is None:
+                        x.v = np.zeros(x.shape)
+                    if x.vhat is None:
+                        x.vhat = np.zeros(x.shape)
+                M = tuple(x.m for x in X)
+                V = tuple(x.v for x in X)
+                Vhat = tuple(x.vhat for x in X)
+
+                proxmin.adaprox(
+                    X,
+                    _grad,
+                    _step,
+                    prox=_prox,
+                    max_iter=max_iter - it,
+                    e_rel=e_rel,
+                    check_convergence=False,
+                    scheme=scheme,
+                    prox_max_iter=prox_max_iter,
+                    callback=callback,
+                    M=M,
+                    V=V,
+                    Vhat=Vhat,
+                    **alg_kwargs
+                )
+
+                logger.info(
+                    "scarlet ran for {0} iterations to logL = {1}".format(
+                        len(self.log_likelihood), self.log_likelihood[-1]
+                    )
+                )
+
+                # set convergence and standard deviation from optimizer
+                for p, m, v, vhat in zip(X, M, V, Vhat):
+                    p.std = 1 / np.sqrt(
+                        ma.masked_equal(v, 0)
+                    )  # this is rough estimate!
+
+                return len(self.log_likelihood), self.log_likelihood[-1]
+
+            # model update forces restart
+            except UpdateException:
```

Comment:
I don't think that this is an issue, since it only does an update once every 10 iterations. But perhaps we should keep track of the iterations where restarts were performed as a diagnostic tool that might be useful later.

Reply:
Not a bad idea, but I won't hold up the PR because of that. Can be done afterwards.